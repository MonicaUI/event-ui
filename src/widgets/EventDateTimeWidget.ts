import { connect } from 'react-redux';
import { getEventTimezone } from '../redux/reducer';
import { adjustTimeZoneTimesForEvent } from 'event-widgets/redux/selectors/event';
import { getSelectedTimezone, getTimeZoneForEventTimeWidgets } from 'event-widgets/redux/selectors/timezone';
import { openTimeZoneDialog } from '../dialogs';
import EventDateTimeWidget from 'event-widgets/lib/EventDateTime/EventDateTimeWidget';

/**
 * Data wrapper for the event date time widget.
 */
export default connect(
  (state: $TSFixMe) => {
    const {
      text: { translate, translateDate, translateTime }
    } = state;
    const eventTimezone = getEventTimezone(state);
    const selectedTimeZone = getSelectedTimezone(state);
    let eventInfo = {
      startDate: state.event.startDate,
      endDate: state.event.endDate
    };
    eventInfo = adjustTimeZoneTimesForEvent(eventInfo, selectedTimeZone, eventTimezone);
    const timeZone = getTimeZoneForEventTimeWidgets(eventTimezone, selectedTimeZone, translate);
    return {
      startDate: eventInfo.startDate,
      endDate: eventInfo.endDate,
      timeZone,
      translateDate,
      translateTime
    };
  },
  {
    switchTimeZone: openTimeZoneDialog
  }
)(EventDateTimeWidget);
