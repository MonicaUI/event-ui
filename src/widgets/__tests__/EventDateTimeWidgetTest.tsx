import React from 'react';
import EventDateTimeWidget from '../EventDateTimeWidget';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

function translateDate(value, format) {
  return `${format}:${value.toISOString().slice(0, 10)}`;
}

function translateTime(value) {
  return value.toISOString().slice(11, 16);
}

function translate(res, opts) {
  return opts ? `${res}:${JSON.stringify(opts)}` : res;
}

const startDate = '2020-06-17T09:00:00.000Z';
const endDate = '2020-06-18T21:00:00.000Z';

const state = {
  appData: {},
  text: {
    translate,
    translateTime,
    translateDate,
    resolver: {}
  },
  event: {
    timezone: 40,
    startDate,
    endDate
  },
  timezones: {
    40: {
      id: 40,
      name: 'US Eastern Time',
      nameResourceKey: 'Event_Timezone_Name_40__resx',
      plannerDisplayName: '(GMT-05:00) Indiana [East]',
      hasDst: false,
      utcOffset: -300,
      abbreviation: 'EST',
      abbreviationResourceKey: 'EST translation',
      dstInfo: []
    }
  }
};

const props = {
  classes: {},
  style: {},
  translate,
  currency: (...args) => `currency:${JSON.stringify(args)}`,
  translateTime,
  translateDate,
  config: {
    displayDateOption: 'dateTimeTimezone'
  }
};

describe('EventDateTimeWidget', () => {
  test('renders', async () => {
    const store = {
      getState() {
        return state;
      },
      subscribe() {},
      dispatch() {}
    };
    const wrapper = renderer
      .create(
        // @ts-expect-error ts-migrate(2769) FIXME: No overload matches this call.
        <Provider store={store}>
          {/* @ts-expect-error ts-migrate(2322) FIXME: Type '{ store: { getState(): { appData: {}; text: ... Remove this comment to see the full error message */}
          <EventDateTimeWidget {...props} store={store} />
        </Provider>
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('EventDateTimeWidget selected timezone tests', () => {
  test('selecting device timezone changes the time and does not append abbreviation', async () => {
    const store = {
      getState() {
        return {
          ...state,
          appData: {
            ...state.appData,
            timeZoneSetting: {
              displayTimeZone: true,
              selectedWidgets: ['eventDateTime']
            }
          },
          selectedTimeZone: {
            utcOffset: 330,
            value: 1001,
            abbreviation: 'IST',
            nameResourceKey: 'your device time zone'
          }
        };
      },
      subscribe() {},
      dispatch() {}
    };
    const getTimezoneOffset = Date.prototype.getTimezoneOffset;
    // eslint-disable-next-line no-extend-native
    Date.prototype.getTimezoneOffset = () => {
      return -330;
    };
    const wrapper = renderer
      .create(
        // @ts-expect-error ts-migrate(2769) FIXME: No overload matches this call.
        <Provider store={store}>
          {/* @ts-expect-error ts-migrate(2322) FIXME: Type '{ store: { getState(): { appData: { timeZone... Remove this comment to see the full error message */}
          <EventDateTimeWidget {...props} store={store} />
        </Provider>
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.children[0].children[1].children[0]).toBe(
      'Widget_DateTimeWidget_TimeRangeText__resx:{"startTime":"19:30","endTime":"07:30"}'
    );
    expect(wrapper.children[0].children[1].children[1].children[0]).toBe('');
    expect(wrapper.children[1].props['data-cvent-id']).toBe('timeZoneWidgetContainer');
    expect(wrapper.children[1].children[0].children[1]).toBe(
      'Event_GuestSide_TimeZone_ViewText__resx:{"timeZone":"your device time zone"}'
    );
    // eslint-disable-next-line no-extend-native
    Date.prototype.getTimezoneOffset = getTimezoneOffset;
  });

  test('selecting event timezone does not change the time and appends abbreviation', async () => {
    const store = {
      getState() {
        return {
          ...state,
          appData: {
            ...state.appData,
            timeZoneSetting: {
              displayTimeZone: true,
              selectedWidgets: ['eventDateTime']
            }
          },
          selectedTimeZone: {
            utcOffset: -300,
            value: 40,
            abbreviation: 'EST',
            abbreviationResourceKey: 'EST translation',
            nameResourceKey: 'Eastern Standard Time'
          }
        };
      },
      subscribe() {},
      dispatch() {}
    };
    const wrapper = renderer
      .create(
        // @ts-expect-error ts-migrate(2769) FIXME: No overload matches this call.
        <Provider store={store}>
          {/* @ts-expect-error ts-migrate(2322) FIXME: Type '{ store: { getState(): { appData: { timeZone... Remove this comment to see the full error message */}
          <EventDateTimeWidget {...props} store={store} />
        </Provider>
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.children[0].children[1].children[0]).toBe(
      'Widget_DateTimeWidget_TimeRangeText__resx:{"startTime":"09:00","endTime":"21:00"}'
    );
    expect(wrapper.children[0].children[1].children[1].children[0]).toBe('EST translation');
    expect(wrapper.children[1].props['data-cvent-id']).toBe('timeZoneWidgetContainer');
    expect(wrapper.children[1].children[0].children[1]).toBe(
      'Event_GuestSide_TimeZone_ViewText__resx:{"timeZone":"Eastern Standard Time"}'
    );
  });

  test('disabling option from site editor hides timezone switch widget', async () => {
    const store = {
      getState() {
        return {
          ...state,
          appData: {
            ...state.appData,
            timeZoneSetting: {
              displayTimeZone: true,
              selectedWidgets: ['agenda']
            }
          },
          selectedTimeZone: {
            utcOffset: -300,
            value: 40,
            abbreviation: 'EST',
            abbreviationResourceKey: 'EST translation',
            nameResourceKey: 'Eastern Standard Time'
          }
        };
      },
      subscribe() {},
      dispatch() {}
    };
    const wrapper = renderer
      .create(
        // @ts-expect-error ts-migrate(2769) FIXME: No overload matches this call.
        <Provider store={store}>
          {/* @ts-expect-error ts-migrate(2322) FIXME: Type '{ store: { getState(): { appData: { timeZone... Remove this comment to see the full error message */}
          <EventDateTimeWidget {...props} store={store} />
        </Provider>
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
