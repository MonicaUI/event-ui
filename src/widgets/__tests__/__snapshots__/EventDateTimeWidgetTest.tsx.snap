// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventDateTimeWidget renders 1`] = `
<div
  style={
    Object {
      "flex": "1 1 auto",
      "pointerEvents": undefined,
    }
  }
>
  <div
    className=""
    style={
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 0,
        "flex": "1 1 auto",
        "paddingBottom": 5,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 15,
        "pointerEvents": "none",
      }
    }
  >
    <div
      className=""
      style={Object {}}
    >
      Widget_DateTimeWidget_DateRangeText__resx:{"startDate":"long:2020-06-17","endDate":"long:2020-06-18"}
    </div>
    <div
      className=""
      style={Object {}}
    >
      Widget_DateTimeWidget_TimeRangeText__resx:{"startTime":"09:00","endTime":"21:00"}
      <span
        className=""
        style={Object {}}
      >
        EST translation
      </span>
    </div>
  </div>
  <div
    data-cvent-id="timeZoneWidgetContainer"
    style={
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 0,
        "flex": "1 1 auto",
        "paddingLeft": 15,
      }
    }
  />
</div>
`;

exports[`EventDateTimeWidget selected timezone tests disabling option from site editor hides timezone switch widget 1`] = `
<div
  style={
    Object {
      "flex": "1 1 auto",
      "pointerEvents": undefined,
    }
  }
>
  <div
    className=""
    style={
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 0,
        "flex": "1 1 auto",
        "paddingBottom": 5,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 15,
        "pointerEvents": "none",
      }
    }
  >
    <div
      className=""
      style={Object {}}
    >
      Widget_DateTimeWidget_DateRangeText__resx:{"startDate":"long:2020-06-17","endDate":"long:2020-06-18"}
    </div>
    <div
      className=""
      style={Object {}}
    >
      Widget_DateTimeWidget_TimeRangeText__resx:{"startTime":"09:00","endTime":"21:00"}
      <span
        className=""
        style={Object {}}
      >
        EST translation
      </span>
    </div>
  </div>
  <div
    data-cvent-id="timeZoneWidgetContainer"
    style={
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 0,
        "flex": "1 1 auto",
        "paddingLeft": 15,
      }
    }
  />
</div>
`;

exports[`EventDateTimeWidget selected timezone tests selecting device timezone changes the time and does not append abbreviation 1`] = `
<div
  style={
    Object {
      "flex": "1 1 auto",
      "pointerEvents": undefined,
    }
  }
>
  <div
    className=""
    style={
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 0,
        "flex": "1 1 auto",
        "paddingBottom": 5,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 15,
        "pointerEvents": "none",
      }
    }
  >
    <div
      className=""
      style={Object {}}
    >
      Widget_DateTimeWidget_DateRangeText__resx:{"startDate":"long:2020-06-17","endDate":"long:2020-06-19"}
    </div>
    <div
      className=""
      style={Object {}}
    >
      Widget_DateTimeWidget_TimeRangeText__resx:{"startTime":"19:30","endTime":"07:30"}
      <span
        className=""
        style={Object {}}
      >
        
      </span>
    </div>
  </div>
  <div
    data-cvent-id="timeZoneWidgetContainer"
    style={
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 0,
        "flex": "1 1 auto",
        "paddingLeft": 15,
      }
    }
  >
    <div
      className=""
      data-cvent-id="timeZoneWidget"
      style={Object {}}
    >
      <span
        className="container"
      >
        <span
          aria-hidden="true"
          className="timeIconSize"
          role="img"
        />
      </span>
      Event_GuestSide_TimeZone_ViewText__resx:{"timeZone":"your device time zone"}
      <span
        className=""
        data-cvent-id="switchButton"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        style={Object {}}
        tabIndex={0}
      >
        <u>
          Event_GuestSide_SwitchText__resx
        </u>
      </span>
    </div>
  </div>
</div>
`;

exports[`EventDateTimeWidget selected timezone tests selecting event timezone does not change the time and appends abbreviation 1`] = `
<div
  style={
    Object {
      "flex": "1 1 auto",
      "pointerEvents": undefined,
    }
  }
>
  <div
    className=""
    style={
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 0,
        "flex": "1 1 auto",
        "paddingBottom": 5,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 15,
        "pointerEvents": "none",
      }
    }
  >
    <div
      className=""
      style={Object {}}
    >
      Widget_DateTimeWidget_DateRangeText__resx:{"startDate":"long:2020-06-17","endDate":"long:2020-06-18"}
    </div>
    <div
      className=""
      style={Object {}}
    >
      Widget_DateTimeWidget_TimeRangeText__resx:{"startTime":"09:00","endTime":"21:00"}
      <span
        className=""
        style={Object {}}
      >
        EST translation
      </span>
    </div>
  </div>
  <div
    data-cvent-id="timeZoneWidgetContainer"
    style={
      Object {
        "backgroundColor": "transparent",
        "borderRadius": 0,
        "flex": "1 1 auto",
        "paddingLeft": 15,
      }
    }
  >
    <div
      className=""
      data-cvent-id="timeZoneWidget"
      style={Object {}}
    >
      <span
        className="container"
      >
        <span
          aria-hidden="true"
          className="timeIconSize"
          role="img"
        />
      </span>
      Event_GuestSide_TimeZone_ViewText__resx:{"timeZone":"Eastern Standard Time"}
      <span
        className=""
        data-cvent-id="switchButton"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        style={Object {}}
        tabIndex={0}
      >
        <u>
          Event_GuestSide_SwitchText__resx
        </u>
      </span>
    </div>
  </div>
</div>
`;
